packages/upstash-redis/src/index.ts

```diff
@@ -0,0 +1,209 @@
+import type { Account as AdapterAccount } from "next-auth"
+import type {
+  Adapter,
+  AdapterUser,
+  AdapterSession,
+  VerificationToken,
+} from "next-auth/adapters"
+import type { Upstash } from "@upstash/redis/src/types"
+import * as uuid from "uuid"
+
+export interface UpstashRedisAdapterOptions {
+  baseKeyPrefix?: string
+  accountKeyPrefix?: string
+  accountByUserIdPrefix?: string
+  emailKeyPrefix?: string
+  sessionKeyPrefix?: string
+  sessionByUserIdKeyPrefix?: string
+  userKeyPrefix?: string
+  verificationTokenKeyPrefix?: string
+}
+
+const defaultOptions = {
+  baseKeyPrefix: "",
+  accountKeyPrefix: "user:account:",
+  accountByUserIdPrefix: "user:account:by-user-id:",
+  emailKeyPrefix: "user:email:",
+  sessionKeyPrefix: "user:session:",
+  sessionByUserIdKeyPrefix: "user:session:by-user-id:",
+  userKeyPrefix: "user:",
+  verificationTokenKeyPrefix: "user:token:",
+}
+
+export default function UpstashRedisAdapter(
+  client: Upstash,
+  options: UpstashRedisAdapterOptions = {}
+): Adapter {
+  const mergedOptions = {
+    ...defaultOptions,
+    ...options,
+  }
+
+  const { baseKeyPrefix } = mergedOptions
+  const accountKeyPrefix = baseKeyPrefix + mergedOptions.accountKeyPrefix
+  const accountByUserIdPrefix =
+    baseKeyPrefix + mergedOptions.accountByUserIdPrefix
+  const emailKeyPrefix = baseKeyPrefix + mergedOptions.emailKeyPrefix
+  const sessionKeyPrefix = baseKeyPrefix + mergedOptions.sessionKeyPrefix
+  const sessionByUserIdKeyPrefix =
+    baseKeyPrefix + mergedOptions.sessionByUserIdKeyPrefix
+  const userKeyPrefix = baseKeyPrefix + mergedOptions.userKeyPrefix
+  const verificationTokenKeyPrefix =
+    baseKeyPrefix + mergedOptions.verificationTokenKeyPrefix
+
+  const reviveFromJson = (json: string) =>
+    JSON.parse(json, (key, value) =>
+      key === "emailVerified" || key === "expires" ? new Date(value) : value
+    )
```

Comment:
I generally don't like it if Date fields are converted by their name. It makes it harder if the user would like to add another Date field. They'll need to modify the adapter.

This seems to be a reoccurring thing with adapters. So far very few adapters respect JS `Date` as-is. One of them is the Prisma adapter.

I think we will need to expose a transform configuration option from adapters. Feel free to ignore this comment for now, it's more of a future self-note.